### PR TITLE
Introduce a cummulative moving average to CKMS

### PR DESCRIPTION
**Pull request by blt**
This commit introduces a cummulative moving average and supporting
QC tests to CKMS. It also includes a fair bit of rehuffling due to
rustfmt but the main deal is really down in src/ckms.

Of particular interest we've had to enforce conversion of T to f64.
This means that some types like i64 are no longer available for use
with the CKMS. Fully generic numeric code is hard in Rust. :/

Signed-off-by: Brian L. Troutwine <blt@postmates.com>

### CODE REVIEW

File: src/ckms.rs [hunk 1]
```diff
@@ -50,11 +51,13 @@ pub struct CKMS<T: Copy> {
     samples: Vec<Entry<T>>,
 
     sum: Option<T>,
+    cma: Option<f64>,
     last_in: Option<T>,
 }
 
 impl<T> AddAssign for CKMS<T>
-    where T: Copy + Add<Output = T> + PartialOrd + Debug
+    where T: Copy + Add<Output = T> + Sub<Output = T> + Div<Output = T> +
```
Comment (pulltab) [inline src/ckms.rs]: Can you enlighten me as to what is happening here?

File: src/ckms.rs [hunk 2]
```diff
@@ -64,13 +67,25 @@ impl<T> AddAssign for CKMS<T>
             (Some(x), None) => Some(x),
             (Some(x), Some(y)) => Some(x.add(y)),
         };
+        self.cma = match (self.cma, rhs.cma) {
```
Comment (pulltab) [inline src/ckms.rs]: Sorry, been awhile.

What is the significance of setting and referencing self.cma here vs say the value of rhs?

Comment (pulltab) [inline src/ckms.rs]: Found my answer a bit further on: https://github.com/postmates/quantiles/pull/13/files#diff-58ef8b5b1c5d3c81aa3b3e58a9da3810R185

File: src/lib.rs [hunk 3]
```diff
@@ -9,7 +9,7 @@
         missing_debug_implementations, missing_copy_implementations,
         unsafe_code,
         unstable_features,
-        unused_import_braces, unused_qualifications)]
```
Comment (pulltab) [inline src/lib.rs]: Why was this removed?

File: src/ckms.rs (hunk 2 above)
Comment (blt) [inline src/ckms.rs]: No worries! 

To start, `cma` is 'cumulative moving average'. If we back up in the definition a little bit the function is `fn add_assign(&mut self, rhs: CKMS<T>)`. We're modifying `self` and consuming `rhs`. What we're doing here in line 70 is  absorbing the value of `rhs.cma` into `self.cma` in the same fashion as sum.

File: src/lib.rs (hunk 3 above)
Comment (blt) [inline src/lib.rs]: The linter around `unused_qualitifications` is buggy. When I added the numeric constraints to `AddAssign for CKMS<T>` – which are necessary to compile – this linter flagged them as unused.

File: src/ckms.rs (hunk 1 above)
Comment (blt) [inline src/ckms.rs]: Sure! `CMKS` is generic on its numeric type `T`. To signal that we need some operations available we gotta add type constraints, which is what's going on here. For instance

    Sub<Output = T>

says that `T` must support the subtraction operation and that `- :: T -> T -> T`.
